3.8.0_to_4.0.0_mysql.sql

```diff
@@ -352,6 +352,69 @@ DROP TABLE `COMPONENTMETRICS`;
 DROP TABLE `CPEREFERENCE`;
 
 
+-- Find dangling components that no project depends on anymore and remove them
+
+DELIMITER $$
+
+CREATE PROCEDURE cleanup_components()
+BEGIN
+    DECLARE v_component_id BIGINT(20);
+    DECLARE v_done BIT DEFAULT FALSE;
+
+    DECLARE component_cursor CURSOR FOR
+        SELECT c.`ID`
+        FROM `COMPONENT` c
+        WHERE (SELECT COUNT(*) FROM `DEPENDENCY` d WHERE c.`ID` = d.`COMPONENT_ID`) = 0;
+    
+    DECLARE CONTINUE HANDLER FOR NOT FOUND SET v_done = TRUE;
+
+    OPEN component_cursor;
+
+    component_loop: LOOP
+        FETCH component_cursor INTO v_component_id;
+
+        IF v_done THEN
+            LEAVE component_loop;
+        END IF;
+
+        IF v_component_id IS NOT NULL THEN
+            DELETE ac 
+            FROM `ANALYSISCOMMENT` ac 
+                LEFT JOIN `ANALYSIS` a ON ac.`ANALYSIS_ID` = a.`ID` 
+            WHERE a.`COMPONENT_ID` = v_component_id;
+
+            DELETE a 
+            FROM `ANALYSIS` a 
+            WHERE a.`COMPONENT_ID` = v_component_id;
```

Comment:
Minor simplification possibility: there is no need to use aliases in single-table statements.

This one could be
```
DELETE FROM `ANALYSIS` WHERE `COMPONENT_ID` = v_component_id;
```